### PR TITLE
More explicit diagnostic when using a `vec![]` in a pattern

### LINKED ISSUE #61933: The `vec![..]` macro can be used in pattern position with bad diagnostics
```
fn main() {
    match Some(vec![3]) {
        Some(vec![x]) => (),
        _ => (),
    }
}
```

Gives error message:

```
error: unexpected `(` after qualified path
 --> src/main.rs:3:14
  |
3 |         Some(vec![x]) => (),
  |              ^^^^^^^
  |              |
  |              unexpected `(` after qualified path
  |              in this macro invocation
  |
  = note: this warning originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

But this is clearly wrong as there is no `(` in the span pointed to.

### PR DESCRIPTION
**Pull request by estebank**
```
error: unexpected `(` after qualified path
  --> $DIR/vec-macro-in-pattern.rs:3:14
   |
LL |         Some(vec![x]) => (),
   |              ^^^^^^^
   |              |
   |              unexpected `(` after qualified path
   |              in this macro invocation
   |              use a slice pattern here instead
   |
   = help: for more information, see https://doc.rust-lang.org/edition-guide/rust-2018/slice-patterns.html
   = note: this warning originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

Fix #61933.

### CODE REVIEW

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 1]
```diff
@@ -70,6 +70,29 @@ impl<'a> ParserAnyMacro<'a> {
             } else if !parser.sess.source_map().span_to_filename(parser.token.span).is_real() {
                 e.span_label(site_span, "in this macro invocation");
             }
+            match kind {
+                AstFragmentKind::Ty => {
```
Comment (Centril) [inline src/libsyntax/ext/tt/macro_rules.rs]: Can you refactor this `match` and the one in `expand.rs` into a function and then make this match into:
```rust
            match kind {
                AstFragmentKind::Pat if macro_ident.name == sym::vec => {
                    e.span_label(
                        site_span,
                        "use a slice pattern here instead",
                    );
                    e.help("for more information, see https://doc.rust-lang.org/edition-guide/\
                              rust-2018/slice-patterns.html");
                }
                _ => foobar(&mut e, kind), // <-- The new function.
            }
```

File: src/libsyntax/ext/tt/macro_rules.rs [hunk 2]
```diff
@@ -70,6 +70,29 @@ impl<'a> ParserAnyMacro<'a> {
...
+                        "this macro call doesn't expand to a type",
+                    );
+                }
+                AstFragmentKind::Pat if macro_ident.name == sym::vec => {
+                    e.span_label(
+                        site_span,
+                        "use a slice pattern here instead",
+                    );
+                    e.help("for more information, see https://doc.rust-lang.org/edition-guide/\
+                              rust-2018/slice-patterns.html");
```
Comment (Centril) [inline src/libsyntax/ext/tt/macro_rules.rs]: Give a suggestion where you have removed `vec!`?

Comment (estebank) [reply src/libsyntax/ext/tt/macro_rules.rs]: I'd prefer to do that in a follow up PR, but yes that's my endgame here.

Comment (Centril) [reply src/libsyntax/ext/tt/macro_rules.rs]: Fair -- file an issue for the follow-up and cc me?

**Comment by Centril** [comment]
r=me if you want with issue filed + refactoring :)

File: src/libsyntax/ext/tt/macro_rules.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/ext/tt/macro_rules.rs]: Done.

**Pull request merged**